### ISSUE TITLE
Automatic Workflows Failing Since February/April

### ISSUE THREAD

**Issue by henningjp** [CoolProp/CoolProp#2394]
### Description

**Automated tests failing for many reasons recently; failing regularly with nightly builds...**
- Release workflow
   - Node.js 16 actions are deprecated.  Updated needed to Node.js 20.  These warnings will start throwing errors after June 30.
   - No artifacts created  
- Development Microsoft C++ Code Analysis
   - Same Node.js error as above
   - SARIF file not processed: mor runs than allowed (46 > 20 maximum); a recent change?

**...and testing of new commits/PRs**
- Development cppcheck 
   - These are all CATCH statements.  Has something changed with Catch2?
   
   ![image](https://github.com/CoolProp/CoolProp/assets/17114032/4eadbabf-5d92-4a71-9295-f8089fef64fb)

- Javascript library
   - em++: error: --memory-init-file is no longer supported.
   See: Emscripten-core/emscripten#21217
   As I think we depend on wasm2js, this could be a can of worms.
- Testing Catch2
   - Some of this is inability to find REFPROP and other libraries in the build environment
   > test cases:    80 |    53 passed |  27 failed
   > assertions: 47130 | 46824 passed | 306 failed

### Versions

**CoolProp Version:** 6.6.1dev

### Additional Information

This is way over my head to even start to tackle.  Is there some documentation on how this is all put together?

This looks like changes to the testing environment, not something that was pushed to the code.  Some of them may be related to the same issue.  The Emscripten switch and the Node.js updates look to be mandatory.  The Catch stuff is a more obscure.

**Comment by ibell** [comment]
Regarding Catch2, did someone update it? They (Catch2) have made quite some changes, but they should not impact us unless somehow has updated the submodule

**Comment by henningjp** [comment]
I don't think so.  looks like it's still pointing to the old version, but I don't know why it's breaking.

**Comment by henningjp** [comment]
Looks like Catch2 is running now, but
- The web structure where Catch2 is running can't fine REFPROP or its FLUIDS/MIXTURES folders
- Some Ancillary checks not agreeing to specified tolerance (was there an update that wasn't reflected in the check values?)
- Humid Air checks not agreeing to specified tolerance (was there an updated that wasn't reflected in the check values?)

**Comment by henningjp** [comment]
@ibell - Can I get to the automation environment to debug the REFPROP missing issue or is that an @jowr only thing?

**Comment by ibell** [comment]
There are mechanisms in GH Actions to have secrets and proprietary things to run actions, but I haven't done so before.

**Comment by ibell** [comment]
@henningjp  Can you fix the Mathcad wrapper? It has compile errors in master

**Comment by henningjp** [comment]
Nothing wrong with the Mathcad wrapper, it compiles perfectly on my local system under VS.

I've been seeing this for a while now on commits.  It uses a header file that comes from PTC to access the Mathcad library.  The errors we're seeing show some error information that some of the structures aren't supported in C++.  However, I know for a fact that this is a Std C header.  So, it is failing.

Interestingly, it does not fail on the nightly builds.  The [nightly builds](https://github.com/CoolProp/CoolProp/actions/runs/10747976528) are compiling the Mathcad Prime wrapper just fine.  Whatever environment we are using for new commits/PRs isn't understanding this is a C header, but the nightly build does.  I don't know how to fix that.  If you wait for the nightly build to run, the Mathcad wrapper will pass.  There are a number of other errors being thrown in the commit/PR test that just don't make sense and all have to do with this header file and the fact that it doesn't get interpreted correctly.

If you wait for the nightly build to run, the Mathcad wrapper will pass. That has been my approach since April when this started happening.

If you have a thought and can point me in the right direction, I'm happy to go poke at it.

**Comment by henningjp** [comment]
It may be a permissions issue with the "hidden" Mathcad component.  I think the runner for commits and PRs runs as the user making the change.  The nightly build runs as the owner of the proprietary actions.  Working on testing that theory.

**Comment by henningjp** [comment]
@ibell, @jowr  - Reading into how this works.  There are two files needed to compile the Mathcad wrapper (one header file and one library) belong to PTC and are not open source.  They have been placed in a protected Private Dependencies directory with specific privileges.  Access is granted through an access key stored in a Repository "Secret".  This allows the runner to use these files in the nightly build.  However, per GitHub:

   > Anyone with collaborator access to this repository can use these secrets and variables for actions. **_<ins>They are not passed to workflows that are triggered by a pull request from a fork</ins>_**.

So, when pushes and pull occur, the necessary header and library can't be retrieved, or they are retrieved in an encrypted format, and the compile fails in odd ways that simply don't make sense.

### So, we've got a few options:

1.  I can try to get permission from PTC to simply post the header and library in the repository under wrappers/Mathcad.  These are fairly static and may not ever need to be updated; the header file hasn't changed in 20 years and I've never had issues cross-compiling with the lib files distributed with different versions of Mathcad Prime.  Then we just copy the file to the runner location without the security measures.   

2. There may be a setting that makes this "Secret" more generally available to other private forks, but I can't explore this without admin access to the repository.   

3. Just disable the Mathcad runner.  Not sure if the Mathcad runner will still get called by the nightly build if we do this.  

4. Alternatively, I'm still researching this, it may be possible to modify the Mathcad runner to check for the Secret token and if unavailable, just abort the runner.  This would allow the runner to build nightly, but skip the build step for pushes and pulls.  I'll test this and see if this is a possibility.

**Comment by henningjp** [comment]
I think there's an easy fix.  Can't at the moment, but will try later this weekend.

**Comment by henningjp** [comment]
@ibell - you probably know this already, but since your pushes on 9/7, the docs_docker is failing and causing the nightly builds to fail.  Nothing pushed to Source Forge since 9/7.

**Comment by henningjp** [comment]
@ibell - I updated the release action from actions/upload-artifact@v3 to actions/upload-artifact@v4.  This should fix the nightly builds, if it doesn't break anything else!  We'll see tonight.

Two issues remaining:

1. **Python cibuildwheel** - failing to upload the sdist tar.gz file to (Test)py.pl.  This step gets skipped on PRs, but not when the PR gets merged into the master branch.  This one is beyond me.  Not sure why we're uploading to (Test)py.pl or if the sdist should be pushed there along with all the wheels (which upload successfully).
2. **Catch2** - Still failing due to:
   - Trying to use the REFPROP backend, which doesn't exist for pushes and PRs
   - Maybe a problem with a UNIFAC backend as well.
   - Many checks missing the target values by a hair (just outside the tolerance set in the check).
   - This one I can take a run at next.

**Comment by ibell** [comment]
For the cibuildwheel, yeah that's basically ok, but I thought it was setup so that it would rename the files to allow for upload to PYPI test. It will run properly if we get around to making a release one of these days.

Don't sweat the Catch2 tests. Obviously it is better to have zero fail so we can see when we break things, but it hasn't been like that for a long time.

**Comment by ibell** [comment]
With the shared libraries, there are some serious non-backwards compatible changes when going to actions/upload-artifact@v4. Unique names are needed for the artifacts, and then they need to be merged after. You can see how I handled that for Python and emulate it

**Comment by henningjp** [comment]
@ibell & @jowr - I think everything is updated in the nightly builds except for one thing.  The mathcad_builder contains a secret that I believe I don't have access to.  Since I pushed the last change to the scheduled release jobs (release_all_files.yaml), they are rescheduled with me as the `actor` and the mathcad_buider is failing.

To remedy, I need Jorrit to **_disable_** and then **_reenable_** the `release_all_files.yaml`.  This will reschedule the workflow with you as the actor and should fix the issue since the secrets used are privately owned (either by you or by admins).  

- Alternatively, this "_might_" work if Ian does this since you are an admin.  We could try that.

> This solution for access to the PTC files needed to compile is not ideal.  I'll see if I can get permission from PTC to just throw them in the repository.  Someone (not me) asked on the Mathcad forum back in 2014 and got no response.  I have some direct contacts and will follow up.

**Comment by henningjp** [comment]
Well, that's not it.  On further inspection, the jobs are being scheduled by Jorrit already.  More tinkering.

**Comment by henningjp** [comment]
@jowr Welcome back!  Can you do me a favor?

1. Touch the two release*.yaml files as well as the mathcad_builder.yaml file so you are the last to edit?
2. Stop and restart the nightly build?

I believe the secret being used to build the mathcad binary isn't being accessed because you set it up.  Something is running as me and failing because I don't have credentials to access the secret and the files come across encrypted.

I have a conversation going with PTC for a better workaround, but going is slow.  I'm hopeful this will get the nightly build going again (including the Mathcad binaries).  Thanks!

**Comment by henningjp** [comment]
Nightly builds running again.
Working on other code tests.
